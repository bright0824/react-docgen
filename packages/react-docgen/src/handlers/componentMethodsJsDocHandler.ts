import parseJsDoc from '../utils/parseJsDoc.js';
import type {
  default as Documentation,
  MethodDescriptor,
} from '../Documentation.js';
import type { Handler } from './index.js';

// Merges two objects ignoring null/undefined.
function merge<T, U>(obj1: T, obj2: U): (T & U) | null {
  if (obj1 == null && obj2 == null) {
    return null;
  }
  const merged: Record<string, unknown> = {
    ...(obj1 as Record<string, unknown>),
  };

  for (const prop in obj2 as Record<string, unknown>) {
    if (obj2[prop] != null) {
      merged[prop] = obj2[prop];
    }
  }

  return merged as T & U;
}
/**
 * Extract info from the methods jsdoc blocks. Must be run after
 * flowComponentMethodsHandler.
 */
const componentMethodsJsDocHandler: Handler = function (
  documentation: Documentation,
): void {
  let methods = documentation.get('methods') as MethodDescriptor[] | null;

  if (!methods) {
    return;
  }

  // @ts-ignore
  methods = methods.map(method => {
    if (!method.docblock) {
      return method;
    }

    const jsDoc = parseJsDoc(method.docblock);

    const returns = merge(jsDoc.returns, method.returns);
    const params = method.params.map(param => {
      const jsDocParam = jsDoc.params.find(p => p.name === param.name);

      return merge(jsDocParam, param);
    });

    return {
      ...method,
      description: jsDoc.description || null,
      returns,
      params,
    };
  });

  documentation.set('methods', methods);
};

export default componentMethodsJsDocHandler;
