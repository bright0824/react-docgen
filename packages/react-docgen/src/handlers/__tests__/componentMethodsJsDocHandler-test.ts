import type { NodePath } from '@babel/traverse';
import DocumentationBuilder from '../../Documentation';
import type { ComponentNode } from '../../resolver';
import type DocumentationMock from '../../__mocks__/Documentation';
import componentMethodsJsDocHandler from '../componentMethodsJsDocHandler.js';
import { beforeEach, describe, expect, test } from 'vitest';

describe('componentMethodsJsDocHandler', () => {
  let documentation: DocumentationBuilder & DocumentationMock;

  beforeEach(() => {
    documentation = new DocumentationBuilder() as DocumentationBuilder &
      DocumentationMock;
  });

  test('stays the same when no docblock is present', () => {
    const methods = [
      {
        name: 'foo',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [
          {
            name: 'test',
            type: null,
          },
        ],
      },
    ];

    documentation.set('methods', methods);
    componentMethodsJsDocHandler(documentation, {} as NodePath<ComponentNode>);
    expect(documentation.get('methods')).toEqual(methods);
  });

  test('adds js doc types when no flow types', () => {
    documentation.set('methods', [
      {
        name: 'foo',
        docblock: `
        @param {string} test
        @returns {string}
      `,
        modifiers: [],
        returns: null,
        params: [
          {
            name: 'test',
            type: null,
          },
        ],
      },
    ]);
    componentMethodsJsDocHandler(documentation, {} as NodePath<ComponentNode>);
    expect(documentation.get('methods')).toMatchSnapshot();
  });

  test('keeps flow types over js doc types', () => {
    documentation.set('methods', [
      {
        name: 'foo',
        docblock: `
        @param {string} test
        @returns {string}
      `,
        modifiers: [],
        returns: {
          type: { name: 'number' },
        },
        params: [
          {
            name: 'test',
            type: { name: 'number' },
          },
        ],
      },
    ]);
    componentMethodsJsDocHandler(documentation, {} as NodePath<ComponentNode>);
    expect(documentation.get('methods')).toMatchSnapshot();
  });

  test('adds descriptions', () => {
    documentation.set('methods', [
      {
        name: 'foo',
        docblock: `
        The foo method.
        @param test The test
        @returns The number
      `,
        modifiers: [],
        returns: null,
        params: [
          {
            name: 'test',
            type: null,
          },
        ],
      },
    ]);
    componentMethodsJsDocHandler(documentation, {} as NodePath<ComponentNode>);
    expect(documentation.get('methods')).toMatchSnapshot();
  });
});
