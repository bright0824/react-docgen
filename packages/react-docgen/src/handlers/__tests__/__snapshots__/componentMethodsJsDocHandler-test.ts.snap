// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componentMethodsJsDocHandler adds descriptions 1`] = `
Array [
  Object {
    "description": "The foo method.",
    "docblock": "
        The foo method.
        @param test The test
        @returns The number
      ",
    "modifiers": Array [],
    "name": "foo",
    "params": Array [
      Object {
        "description": "The test",
        "name": "test",
        "optional": false,
        "type": null,
      },
    ],
    "returns": Object {
      "description": "The number",
      "type": null,
    },
  },
]
`;

exports[`componentMethodsJsDocHandler adds js doc types when no flow types 1`] = `
Array [
  Object {
    "description": null,
    "docblock": "
        @param {string} test
        @returns {string}
      ",
    "modifiers": Array [],
    "name": "foo",
    "params": Array [
      Object {
        "description": null,
        "name": "test",
        "optional": false,
        "type": Object {
          "name": "string",
        },
      },
    ],
    "returns": Object {
      "description": null,
      "type": Object {
        "name": "string",
      },
    },
  },
]
`;

exports[`componentMethodsJsDocHandler keeps flow types over js doc types 1`] = `
Array [
  Object {
    "description": null,
    "docblock": "
        @param {string} test
        @returns {string}
      ",
    "modifiers": Array [],
    "name": "foo",
    "params": Array [
      Object {
        "description": null,
        "name": "test",
        "optional": false,
        "type": Object {
          "name": "number",
        },
      },
    ],
    "returns": Object {
      "description": null,
      "type": Object {
        "name": "number",
      },
    },
  },
]
`;
